R/devel.R

```diff
@@ -17,22 +17,33 @@ dcf.repo = function(pkg, repo, field, type){
   dcf[dcf[,"Package"]==pkg, field][[1L]]
 }
 
-update.dev.pkg = function(object="data.table", repo="https://Rdatatable.gitlab.io/data.table", field="Revision", type=getOption("pkgType"), lib=NULL, ...){
+update.dev.pkg = function(object="data.table", repo="https://Rdatatable.gitlab.io/data.table", field="Revision", type=getOption("pkgType"), lib=NULL, ...) {
+  # this works for any package, not just data.table
   pkg = object
   # perform package upgrade when new Revision present
   stopifnot(is.character(pkg), length(pkg)==1L, !is.na(pkg),
             is.character(repo), length(repo)==1L, !is.na(repo),
             is.character(field), length(field)==1L, !is.na(field),
             is.null(lib) || (is.character(lib) && length(lib)==1L && !is.na(lib)))
+  # get Revision field from remote repository PACKAGES file
   una = is.na(ups<-dcf.repo(pkg, repo, field, type))
-  upg = una | !identical(ups, dcf.lib(pkg, field, lib.loc=lib))
-  if (upg) utils::install.packages(pkg, repos=repo, type=type, lib=lib, ...)
-  if (una) cat(sprintf("No commit information found in DESCRIPTION file for %s package. Unsure '%s' is correct field name in PACKAGES file in your devel repository '%s'.\n", pkg, field, file.path(repo, "src","contrib","PACKAGES")))
-  cat(sprintf("R %s package %s %s (%s)\n",
-              pkg,
-              c("is up-to-date at","has been updated to")[upg+1L],
-              dcf.lib(pkg, field, lib.loc=lib),
-              utils::packageVersion(pkg, lib.loc=lib)))
+  if (una)
+    cat(sprintf("No revision information found in DESCRIPTION file for %s package. Unsure '%s' is correct field in PACKAGES file in your package repository '%s'. Otherwise package will be re-installed every time, proceeding to installation.\n",
+                pkg, field, contrib.url(repo, type=type)))
+  # see if Revision is different then currently installed Revision, note that installed package will have Revision info only when it was installed from remote devel repo
+  upg = una || !identical(ups, dcf.lib(pkg, field, lib.loc=lib))
+  # update.dev.pkg fails on windows R 4.0.0, we have to unload package namespace before installing new version #4403
+  on.exit({
+    if (upg) {
+      unloadNamespace(pkg)
+      utils::install.packages(pkg, repos=repo, type=type, lib=lib, ...)
```

Comment:
Maybe wrap this with a tryCatch()?  Just so that a more helpful error message can be produced instead; e.g. "Please close all R sessions to release the dll lock on Windows, and try again. If the dll lock is still open, try a reboot which should release the lock on Windows. Also, if you have a .Rpofile which loads data.table on startup, turn that off temporarily in order to upgrade the dll on Windows."

Reply:
.Rprofile should not be an issue, because DT is already loaded at this point.

Reply:
I would lean towards keeping `on.exit` chunk as simple as possible